### DISCUSSION TITLE
Configurable preprocess (transform) of values

### DISCUSSION THREAD

**Discussion by aaronadamsCA** [edmundhung/conform#387] [Ideas]
I'd like to see the `@conform-to/zod` form data parser accept some sort of configuration where I can pass in my own `transform` and/or `preprocess` functions to replace the value coercion that is currently built in.

For example, I'd prefer to `preprocess` empty strings and empty files to `undefined` so my schemas could simply use `.required()` checks for these types.

**Comment by edmundhung** [comment]
Conform strips the empty string and does type coercion by introspecting the complete schema and inject extra preprocess step on each field:

```tsx
// Let's say this is the schema you have:
const schema = z.object({
  number: z.number(),
});

// Conform turns it to this:
const schema = z.object({
  number: z.preprocess(value => {
    if (typeof value !== 'string') {
      // Return the value as is if the value is not string
      return value;
    }

    if (value === '') {
      // Strip empty string
      return undefined;
    }

    // Type coercion
    return Number(value);
  }, z.number()),
});

// If you add your own preprocess in the schema, like this:
const schema = z.object({
  number: z.preprocess(value => {
    // your custom preporcess logic
  }, z.number()),
});

// Then it will become this:
const schema = z.object({
  number: z.preprocess(value => {
    // your custom preporcess logic
  }, z.preprocess(value => {
    if (typeof value !== 'string') {
      // Return the value as is if the value is not string
      return value;
    }

    if (value === '') {
      // Strip empty string
      return undefined;
    }

    // Type coercion
    return Number(value);
  }, z.number()),
});
```

As your preprocess will run first, you should be able to do whatever you want.

**Reply by edmundhung** [reply]
If you try that and find it not working, it's very likely because you are using zod 3.22 which has a regression with `z.preprocess()`. Please downgrade to 3.21.4 and try again 🙌🏼

**Reply by aaronadamsCA** [reply]
Okay, I think I understand now, thank you for explaining! I could only see usage of `transform` [in the source code](https://github.com/edmundhung/conform/blob/68bd144fb4ba345aace960be4f8b0570df37c287/packages/conform-zod/coercion.ts), and that it was changing our types; but now I can see and understand that your usage of it mimics `preprocess`, and that I should be able to both preprocess and transform it myself after the fact. I'll give this a try.

**Reply by edmundhung** [reply]
Yes. That's because the same reason I mentioned above: some people upgraded to zod 3.22 and find it totally broken. So I have to switch the implementation to get around it 🙃 If you check the [history](https://github.com/edmundhung/conform/pull/283/files), you will see the previous implementation is exactly what I am showing above.

**Comment by aaronadamsCA** [comment]
> For example, I'd prefer to `preprocess` empty strings and empty files to `undefined` so my schemas could simply use `.required()` checks for these types.

Haha, Conform already does this, and `.required()` checks already work. Good to know!

This also means that this works to transform empty strings/files to `null` instead of `undefined`:

```ts
string: z.string().nullable().default(null),
file: z.instanceof(File).nullable().default(null),
```

This all ended up being much easier than I expected!

**Reply by edmundhung** [reply]
This is probably worth a tips on the documentation [here](https://github.com/edmundhung/conform/blob/v1/docs/api/zod/parseWithZod.md). PR welcome 🙌🏼
